esp/esdllib/esdl_script.cpp

```diff
@@ -0,0 +1,919 @@
+}
+
+class CEsdlTransformOperationBase : public CInterfaceOf<IEsdlTransformOperation>
+{
+protected:
+    StringAttr mergedTarget;
+    StringAttr tagname;
+    bool ignoreCodingErrors = false; //ideally used only for debugging
+
+public:
+    CEsdlTransformOperationBase(IPropertyTree *tree, const StringBuffer &prefix)
+    {
+        tagname.set(tree->queryName());
+        if (tree->hasProp("@_crtTarget"))
+            mergedTarget.set(tree->queryProp("@_crtTarget"));
+        ignoreCodingErrors = tree->getPropBool("@optional", false);
+    }
+
+    virtual const char *queryMergedTarget() override
+    {
+        return mergedTarget;
+    }
+};
+
+
+class CEsdlTransformOperationVariable : public CEsdlTransformOperationBase
+{
+protected:
+    StringAttr m_name;
+    Owned<ICompiledXpath> m_select;
+
+public:
+    CEsdlTransformOperationVariable(IPropertyTree *tree, const StringBuffer &prefix) : CEsdlTransformOperationBase(tree, prefix)
+    {
+        m_name.set(tree->queryProp("@name"));
+        if (m_name.isEmpty())
+            esdlOperationError(ESDL_SCRIPT_MissingOperationAttr, tagname, "without name", m_name, !ignoreCodingErrors);
+        const char *select = tree->queryProp("@select");
+        if (!isEmptyString(select))
+            m_select.setown(compileXpath(select));
+    }
+
+    virtual ~CEsdlTransformOperationVariable()
+    {
+    }
+
+    virtual bool process(IEspContext * context, IPropertyTree *content, IXpathContext * xpathContext) override
+    {
+        if (m_select)
+            return xpathContext->addCompiledVariable(m_name, m_select);
+        return xpathContext->addVariable(m_name, "");
+    }
+
+    virtual void toDBGLog() override
+    {
+#if defined(_DEBUG)
+        DBGLOG(">%s> %s with select(%s)", m_name.str(), tagname.str(), m_select.get() ? m_select->getXpath() : "");
+#endif
+    }
+};
+
+class CEsdlTransformOperationParameter : public CEsdlTransformOperationVariable
+{
+public:
+    CEsdlTransformOperationParameter(IPropertyTree *tree, const StringBuffer &prefix) : CEsdlTransformOperationVariable(tree, prefix)
+    {
+    }
+
+    virtual ~CEsdlTransformOperationParameter()
+    {
+    }
+
+    virtual bool process(IEspContext * context, IPropertyTree *content, IXpathContext * xpathContext) override
+    {
+        if (m_select)
+            return xpathContext->declareCompiledParameter(m_name, m_select);
+        return xpathContext->declareParameter(m_name, "");
+    }
+};
+
+class CEsdlTransformOperationSetValue : public CEsdlTransformOperationBase
+{
+protected:
+    Owned<ICompiledXpath> m_select;
+    Owned<ICompiledXpath> m_xpath_target;
+    StringAttr m_target;
+    StringAttr m_traceName;
+
+public:
+    CEsdlTransformOperationSetValue(IPropertyTree *tree, const StringBuffer &prefix) : CEsdlTransformOperationBase(tree, prefix)
+    {
+        m_traceName.set(tree->queryProp("@name"));
+
+        if (isEmptyString(tree->queryProp("@target")) && isEmptyString(tree->queryProp("@xpath_target")))
```

Comment:
Where is @xpath_target defined? Is it a deprecated synonym for @target?

Reply:
No actually xpath_target is a new feature I've added.  By allowing the target to be built as an xpath you can build much more dynamic target paths... the for-each test case shows this.  For each entry I'm able to build a path to it's exact matching target location.  String literals would be hard coded when the script is written.

Reply:
OK I'll refer to the test case.